src/main/java/makaroshyna/onlinebookstore/repository/impl/BookRepositoryImpl.java

```diff
@@ -16,32 +18,37 @@ public class BookRepositoryImpl implements BookRepository {
 
     @Override
     public Book save(Book book) {
-        EntityManager entityManager = null;
         EntityTransaction transaction = null;
-        try {
-            entityManager = entityManagerFactory.createEntityManager();
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
             transaction = entityManager.getTransaction();
+            transaction.begin();
             entityManager.persist(book);
             transaction.commit();
             return book;
         } catch (Exception e) {
-            if (transaction != null) {
+            if (transaction != null && transaction.isActive()) {
                 transaction.rollback();
             }
-            throw new RuntimeException("Can't insert book into DB: " + book, e);
-        } finally {
-            if (entityManager != null) {
-                entityManager.close();
-            }
+            throw new EntityNotFoundException("Can't insert book into DB: " + book, e);
+        }
+    }
+
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            Book book = entityManager.find(Book.class, id);
+            return Optional.ofNullable(book);
+        } catch (Exception e) {
+            throw new EntityNotFoundException("Can't find book with id: " + id, e);
         }
     }
 
     @Override
     public List<Book> findAll() {
         try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
-            return entityManager.createQuery("SELECT b FROM Book b", Book.class).getResultList();
+            return entityManager.createQuery("FROM Book", Book.class).getResultList();
         } catch (Exception e) {
-            throw new RuntimeException("Can't find all books", e);
+            throw new EntityNotFoundException("Can't find all books", e);
```

Comment:
`EntityNotFoundException` is meant more for a service layer error handling, when you check the optional that you get, because if you catch some exception here it won't mean that the entity does not exist, it'll mean that we encountered some other problem like SQL exception or something like that, so let's check it's usage in other places as well.
If you want you can add some other custom exception to throw it in the repository level, like we had `DataProcessingException` before